networkx/algorithms/community/partition.py

```diff
@@ -0,0 +1,276 @@
+# partition.py - functions for partitioning a graph
+#
+# Copyright 2011 Ben Edwards <bedwards@cs.unm.edu>.
+# Copyright 2011 Aric Hagberg <hagberg@lanl.gov>.
+# Copyright 2015 NetworkX developers.
+#
+# This file is part of NetworkX.
+#
+# NetworkX is distributed under a BSD license; see LICENSE.txt for more
+# information.
+"""Functions for partitioning a graph."""
+from __future__ import division
+
+from itertools import chain
+from copy import deepcopy
+
+try:
+    import numpy as np
+except ImportError:
+    is_numpy_available = False
+else:
+    is_numpy_available = True
+
+import networkx as nx
+from networkx.utils import not_implemented_for
+from .community_utils import is_partition
+from .modularity import modularity
+
+__all__ = ['spectral_modularity_partition', 'greedy_max_modularity_partition']
+
+chaini = chain.from_iterable
+
+
+def spectral_modularity_partition(G):
+    r"""Returns a bipartition of the nodes based on the spectrum of the
+    modularity matrix of the graph.
+
+    This method calculates the eigenvector associated with the second
+    largest eigenvalue of the modularity matrix, where the modularity
+    matrix *B* is defined by
+
+    ..math::
+
+        B_{i j} = A_{i j} - \frac{k_i k_j}{2 m},
+
+    where *A* is the adjacency matrix, `k_i` is the degree of node *i*,
+    and *m* is the number of edges in the graph. Nodes whose
+    corresponding values in the eigenvector are negative are placed in
+    one block, nodes whose values are nonnegative are placed in another
+    block.
+
+    Parameters
+    ----------
+    G : NetworkX Graph
+
+    Returns
+    --------
+    C : tuple
+        Pair of sets of nodes of ``G``, partitioned according to second
+        largest eigenvalue of the modularity matrix.
+
+    Raises
+    ------
+    ImportError
+        If ``NumPy`` is not available.
+
+    Examples
+    --------
+    >>> G = nx.karate_club_graph()
+    >>> left, right = nx.spectral_modularity_partition(G)
+    >>> sorted(left)
+    [0, 1, 2, 3, 4, 5, 6, 7, 10, 11, 12, 13, 16, 17, 19, 21]
+    >>> sorted(right)
+    [8, 9, 14, 15, 18, 20, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33]
+
+    Notes
+    -----
+    Defined for instances of the classes :class:`~networkx.Graph`,
+    :class:`~networkx.DiGraph`, :class:`~networkx.MultiGraph`, and
+    :class:`~networkx.MultiDiGraph`.
+
+    References
+    ----------
+    .. [1] M. E. J. Newman *Networks: An Introduction*, pages 373--378
+       Oxford University Press 2011.
+
+    """
+    if not is_numpy_available:
+        raise ImportError('failed to import NumPy: http://www.numpy.org/')
+    B = nx.modularity_matrix(G)
+    eigenvalues, eigenvectors = np.linalg.eig(B)
+    # Get the index of the largest eigenvalue.
+    #
+    # TODO Shouldn't this be -2?
+    index = np.argsort(eigenvalues)[-1]
```

Comment:
In the original code the index was `-1`. However, the second largest eigenvalue should be at index `-2`, right?